parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java

```diff
@@ -677,10 +758,22 @@ private TProtocol getProtocol(ThriftField field, ColumnIO columnIO, Events retur
       p = new MapWriteProtocol((GroupColumnIO)columnIO, (MapType)type, returnClause);
       break;
     case SET:
-      p = new ListWriteProtocol((GroupColumnIO)columnIO, ((SetType)type).getValues(), returnClause);
+      if (writeThreeLevelList) {
```

Comment:
we could refactor this out into a method `createListWriteProtocol(writeLevel3, columnIO, type, returnClause)` that way we don't have this largely identical repeated block